### PR TITLE
Remote Runner / PostMessge api

### PR DESCRIPTION
**Pull request by flashdesignory**
This pr introduces Remote Runner workloads. 
Workloads that are local OR that point to a remote url (when running Speedometer locally), can communicate and run their own tests via PostMessage.

Basic flow:

1. SuiteRunner loads the iframe of the local or remote workload.
2. SuiteRunner waits for the "app-ready" message from the workload.
3. Workload sends "app-ready" message, once its able to listen and receive a command to start the tests.
4. SuiteRunner sends the "benchmark-suite" command to start the testing (with an optional suite name).
5. Workload runs through the test suite and collects performance metrics.
6. After each completed test step it sends a "step-complete" message, to update progress in speedometer.
7. With the "suite-complete" message, the workload sends the metrics along.
8. Speedometer receives metrics.

image

Notes
- TestRunner and TestInvoker files are installed as dependencies in the news-site-next workload, to ensure they stay synced with updates.

### CODE REVIEW

File: resources/newssite/news-next/public/test-invoker.mjs
Comment (bgrins) [inline resources/newssite/news-next/public/test-invoker.mjs:1]: I was surprised this is in public/ since you are loading this file via npm (I would have expected it to only exist in dist/). Is this left over or is this how npm works?

File: resources/newssite/news-next/public/test-runner.mjs
Comment (bgrins) [inline resources/newssite/news-next/public/test-runner.mjs:1]: Ditto

File: resources/newssite/news-next/public/test-invoker.mjs
Comment (flashdesignory) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: yeah, so basically I'm doing this:

1. install dependency through [package.json](https://github.com/WebKit/Speedometer/blob/ee3ceb4c4598ae7517d6a403e99cefdb56a38a53/resources/newssite/news-next/package.json#L24) (this copies the two files in my node_modules folder)
2. run a [script](https://github.com/WebKit/Speedometer/blob/ee3ceb4c4598ae7517d6a403e99cefdb56a38a53/resources/newssite/news-next/scripts/prepare.js#L39) to copy both files to the public folder

The reason is that the code needs to execute in the hosted directory when you visit the workload. 
The node_modules folder is not available at that point.

Comment (bgrins) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: if you were to update the `<script>` src in the development source to point to the full node_modules path wouldn't the build step be able to figure it out to get it into dist without the extra script to copy the file? Not sure how the build works exactly but that's what I figured would happen conceptually

Comment (flashdesignory) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: let me take a look if I can simplify....

Comment (flashdesignory) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: @bgrins - looking at the code, the files I am referencing in the `script` component are all files created by the news-site-workload:

- benchmark-connector.mjs
- workload-test.mjs
- workload-testing-utils.mjs

The last file imports the test-runner, but since I am using the script component, I have to put the 3 files above in the public folder. This is because of the way Next.js resolves file paths.

Second problem: I can't reference an import from the node_modules folder, since the three files above are in the public folder, which doesn't have access to the node_modules.

Comment (bgrins) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: As a test, I ran `npm create vite@latest` and then installed the entire "speedometer" package rather than one-by-one mjs, then referenced it as `import { TestRunner } from 'speedometer/resources/test-runner.mjs'` and it seems to work in both dev and build (see screenshot). I assume with this approach we could also move workload-testing-utils.mjs into the main resources dir instead of implementing it directly in the workload.

I'm not quite sure how Next `<Script>` elements integrate with this, but presumably they have a way to reference something from an installed package (maybe make the script element reference a local file and then have that file ESM import the relevant helpers). Or maybe I'm missing something?

Screenshot 2024-11-21 at 10 33 55 AM

Comment (bgrins) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: #459 shows an example of this

Comment (flashdesignory) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: we need a solution that works for various frameworks though. 
For Next.js, I can't point the Script component to the node_modules folder... 
for example below won't work

```
<Script src="node_modules/benchmark-connector.mjs"...>
```

Comment (flashdesignory) [reply resources/newssite/news-next/public/test-invoker.mjs:1]: > #459 shows an example of this

yeah imports are different from static js files though.

File: resources/newssite/news-next/dist/benchmark-connector.mjs
Comment (bgrins) [inline resources/newssite/news-next/dist/benchmark-connector.mjs:8]: @flashdesignory I don't see the following logic used in the main runner. Since it also reads params from the URL, would it be possible to share some code here instead?

Comment (flashdesignory) [reply resources/newssite/news-next/dist/benchmark-connector.mjs]: we could - the params.mjs is slightly different, but I can take a look at both and see if I can clean up and reuse

File: resources/newssite/news-next/public/test-invoker.mjs
Comment (flashdesignory) [reply resources/newssite/news-next/public/test-invoker.mjs]: @bgrins was convincing 😄 .... refactored and introduced a shared folder that's installable.

**Comment by bgrins** [comment]
Now that this has mostly incorporated #459 (I'd still like to see the params sharing) I think it's going to be easier to review & hopefully get landed for testing @rniwa @julienw

File: resources/newssite/news-next/src/workload-test.mjs
Comment (bgrins) [inline resources/newssite/news-next/src/workload-test.mjs:4]: I guess we could actually just pass this object as a parameter into the call to `connectToBenchmark` now that we're just calling the function (instead of sticking it on the window object). So just have this module call connectToBenchmark directly instead of sticking it on benchmarkSuitesManager and then having the App object call both. WDTY @flashdesignory ?

Comment (flashdesignory) [reply resources/newssite/news-next/src/workload-test.mjs]: done!

File: resources/shared/workload-testing-utils.mjs
Comment (flashdesignory) [inline resources/shared/workload-testing-utils.mjs:143]: for SSR compatibility!

File: resources/shared/workload-testing-utils.mjs
Comment (julienw) [inline resources/shared/workload-testing-utils.mjs:148]: nit: add a default clause with a console.error line writing the received `event.data.type`

File: resources/shared/workload-testing-utils.mjs
Comment (julienw) [inline resources/shared/workload-testing-utils.mjs:130]: is it an error that we don't have name and version? Should we throw or at least output a warning?

File: resources/shared/workload-testing-utils.mjs
Comment (julienw) [inline resources/shared/workload-testing-utils.mjs:123]:
```suggestion
 * When the app is ready, an 'app-ready' message is sent to signal that the app can receive instructions.
```

File: resources/shared/workload-testing-utils.mjs
Comment (julienw) [inline resources/shared/workload-testing-utils.mjs:92]: could the Page class use them directly (maybe as a follow-up, I don't know)
I don't feel comfortable with the duplication.

File: resources/suite-runner.mjs
Comment (julienw) [inline resources/suite-runner.mjs:154]:
```suggestion
        const appReadyPromise = this._subscribeOnce("app-ready");
```

File: resources/suite-runner.mjs
Comment (julienw) [inline resources/suite-runner.mjs:176]: why are we overriding the `tests` property but incrementing the `total` value?
Should we append the `tests` instead?
(I haven't researched closely but it would be good to comment about that)

File: resources/shared/params.mjs
Comment (julienw) [inline resources/shared/params.mjs:149]: can we have an option object parameter instead of a boolean? it's really not clear what this does from the caller site.

(and actually from here neither, a comment would be nice)

File: resources/shared/params.mjs
Comment (julienw) [inline resources/shared/params.mjs:163]: when is it happening that window isn't present?

optional nit: use `?.` operator

**Review by julienw — APPROVED** [review]
I haven't tried locally but this looks pretty good to me. We can always fix things later if there's anything broken anyway.

r+ with my comments fixed.

Note: it would be a lot easier to review if you'd reworked the PR to have consistent commits, for example:

Commit 1: move files around without changing them
Commit 2: add the remote runner + the workload tools
Commit 3: change the workload itself

If you do this for future complex PR this will make them a joy to review (and hence faster to review as well). Thanks!

File: resources/newssite/news-next/src/pages/index.js
Comment (flashdesignory) [reply resources/newssite/news-next/src/pages/index.js:12]: Added a comment. Basically, we want to wait until the workload is ready to accept a test suite. 
Waiting until all DOM mutations are done ensures that elements are present that are needed.
The Workload should decide when the 'app-ready' message should be sent to the benchmark.

File: resources/newssite/news-next/src/pages/index.js
Comment (flashdesignory) [reply resources/newssite/news-next/src/pages/index.js]: added a cleanup function. It's less important here, but I am now also disconnecting after the "suite-complete' message has been sent.

File: resources/suite-runner.mjs
Comment (flashdesignory) [reply resources/suite-runner.mjs:49]: I feel like this should be a discussion point outside of the current prs.

File: resources/newssite/news-next/src/pages/index.js
Comment (julienw) [reply resources/newssite/news-next/src/pages/index.js:12]: but why useEffect and not useLayoutEffect?

File: resources/shared/workload-testing-utils.mjs
Comment (flashdesignory) [reply resources/shared/workload-testing-utils.mjs:92]: Agreed - This should be a follow up!

File: resources/suite-runner.mjs
Comment (flashdesignory) [reply resources/suite-runner.mjs]: done

File: resources/suite-runner.mjs
Comment (flashdesignory) [reply resources/suite-runner.mjs]: cleaned this up - thank you!

File: resources/shared/params.mjs
Comment (flashdesignory) [reply resources/shared/params.mjs:149]: I added some comments in that function.
We don't use option params in our current code base and it feels too much, just for readability. 
We can rename "forRemote" or I can also add a comment on the caller site, if that helps.

Comment (julienw) [reply resources/shared/params.mjs:149]: a comment on the caller site would be good, thanks

File: resources/newssite/news-next/src/pages/index.js
Comment (flashdesignory) [reply resources/newssite/news-next/src/pages/index.js]: You're right with questioning me 😄 - switched to useEffect since that waits for the paint.

**Comment by camillobruni** [comment]
Still looking good from my side.

File: resources/shared/params.mjs
Comment (rniwa) [inline resources/shared/params.mjs:157]: I'd prefer having an allowed list of params instead of letting every param not listed here pass through.

File: resources/shared/workload-testing-utils.mjs
Comment (rniwa) [inline resources/shared/workload-testing-utils.mjs:9]: "Test" doesn't seem to add much value here. Why not just `BenchmarkStep`?

File: resources/shared/workload-testing-utils.mjs
Comment (rniwa) [inline resources/shared/workload-testing-utils.mjs:27]: Ditto for BenchmarkSuite.

File: resources/shared/workload-testing-utils.mjs
Comment (rniwa) [inline resources/shared/workload-testing-utils.mjs:77]: Why do we need this? It doesn't seem like we're gonna have more than one suite per document?

File: resources/newssite/news-next/src/pages/index.js
Comment (rniwa) [inline resources/newssite/news-next/src/pages/index.js:7]: So a bunch of tests are all going to have their own getBenchmarkSuitesManager?
That would make grep'ing harder. Why not name this after each suite?
e.g. `import { NewsNextSuite } from "@/workload-test.mjs";`

File: resources/shared/workload-testing-utils.mjs
Comment (rniwa) [inline resources/shared/workload-testing-utils.mjs:130]: I'd rather have this as a class like `BenchmarkConnector` or maybe `BenchmarkRunnerProxy`.

File: resources/shared/workload-testing-utils.mjs
Comment (rniwa) [inline resources/shared/workload-testing-utils.mjs:89]: I'd rather have a separate file for these helper methods and rename this file to "benchmark.mjs" or something since all classes here will have Benchmark prefix then.

File: resources/suite-runner.mjs
Comment (rniwa) [inline resources/suite-runner.mjs:189]: Can we call get here and use a local variable?
```js
if (count callback = this.postMessageCallbacks.get(e.data.type))
    callback(e);
```

File: resources/suite-runner.mjs
Comment (rniwa) [inline resources/suite-runner.mjs:188]: Please don't use a single letter variable name like "e". Please spell out event.

File: resources/tests.mjs
Comment (rniwa) [inline resources/tests.mjs:906]: Hm... so whenever we update benchmark suite, we need to remember to update this magic number?
That doesn't sound like a scalable solution.

Can we load `resources/newssite/news-next/src/workload-test.mjs` in the harness as well and get the number of suites from there?

Alternatively, can we define some kind of JSON file somewhere which outlines each step of suite?

File: resources/shared/workload-testing-utils.mjs
Comment (flashdesignory) [reply resources/shared/workload-testing-utils.mjs:77]: There is a reason for this. 
Each workload has a 'default' suite, the one that runs by default in Speedometer. 

This setup allows for one workload to have multiple suites. one default and let's say an experimental one. 

It lets us reuse the same workload, with different suites to run.

File: resources/newssite/news-next/src/pages/index.js
Comment (flashdesignory) [reply resources/newssite/news-next/src/pages/index.js:7]: A workload can have multiple suites, but we could name it after the workload.
What's the benefit of renaming it?

File: resources/shared/params.mjs
Comment (flashdesignory) [reply resources/shared/params.mjs:157]: An allowlist is something we'd need to maintain, vs here we just remove what breaks the remote workload.

File: resources/shared/workload-testing-utils.mjs
Comment (flashdesignory) [reply resources/shared/workload-testing-utils.mjs]: done

File: resources/shared/workload-testing-utils.mjs
Comment (flashdesignory) [reply resources/shared/workload-testing-utils.mjs]: done

File: resources/tests.mjs
Comment (camillobruni) [reply resources/tests.mjs:906]: An alternative solution we discussed is a little shared helper library that communicates back the progress / step sizes on Test setup.
Given that we don't really have many workloads yet I'd prefer sticking to slightly more cumbersome manual approach and work towards a more maintainable solution in the long run.

File: resources/suite-runner.mjs
Comment (camillobruni) [reply resources/suite-runner.mjs:49]: I tend to be on your side here julien... but I totally fine with cleaning / discussing this after landing this PR :)

File: resources/suite-runner.mjs
Comment (camillobruni) [inline resources/suite-runner.mjs:133]: this can definitely be just a public field if we have public getters *and* setters.

File: resources/tests.mjs
Comment (bgrins) [reply resources/tests.mjs:906]: +1 on doing this as a followup. We can discuss there whether JSON manifest with hardcoded step count or encoding it in the setup message would be preferable.

File: resources/shared/params.mjs
Comment (flashdesignory) [inline resources/shared/params.mjs:111]: oversight?

File: resources/shared/params.mjs
Comment (flashdesignory) [reply resources/shared/params.mjs]: we fixed the params obj without filtering keys.

File: resources/suite-runner.mjs
Comment (flashdesignory) [reply resources/suite-runner.mjs]: implemented above

File: resources/shared/workload-testing-utils.mjs
Comment (flashdesignory) [reply resources/shared/workload-testing-utils.mjs]: as discussed, using a 'suites' object now.

File: resources/newssite/news-next/src/pages/index.js
Comment (flashdesignory) [reply resources/newssite/news-next/src/pages/index.js]: with the mentioned suites object, which is exported as default, we could easily rename the import if that's desired.

File: resources/tests.mjs
Comment (flashdesignory) [reply resources/tests.mjs]: removed steps magic number!

File: resources/newssite/news-next/src/workload-test.mjs
Comment (rniwa) [inline resources/newssite/news-next/src/workload-test.mjs:1]: Don't we need to import from benchmark.mjs and helpers.mjs now that they're split?

File: resources/shared/benchmark.mjs
Comment (rniwa) [inline resources/shared/benchmark.mjs:18]: Do we need an explicit await here? Can't we just return the result of `testRunner.runTest()`?

File: resources/newssite/news-next/src/workload-test.mjs
Comment (flashdesignory) [reply resources/newssite/news-next/src/workload-test.mjs:1]: the imported files re-exports them

File: resources/suite-runner.mjs
Comment (rniwa) [inline resources/suite-runner.mjs:26]: Do we really need to expose all these methods now?
It doesn't seem like this is ever called anywhere??
Ditto for other methods in this class.

**Review by rniwa — APPROVED** [review]

File: resources/newssite/news-next/src/workload-test.mjs
Comment (rniwa) [reply resources/newssite/news-next/src/workload-test.mjs:1]: Do we really need that? Directly importing from benchmark.mjs & helpers.mjs seem better?

File: resources/shared/benchmark.mjs
Comment (flashdesignory) [reply resources/shared/benchmark.mjs:18]: yes, since the RAFTestInvoker has this line:

`
const result = await this._reportCallback();
`

File: resources/suite-runner.mjs
Comment (flashdesignory) [reply resources/suite-runner.mjs:26]: yeah, since the RemoteSuiteRunner extends the SuiteRunner and the private vars are not available in the RemoteSuiteRunner.

I have to use getters to be able to reference the frame like:
`this.frame.contentWindow.postMessage`

Comment (rniwa) [reply resources/suite-runner.mjs:26]: Oh, I missed that these are getters, not methods.

File: resources/newssite/news-next/src/workload-test.mjs
Comment (flashdesignory) [reply resources/newssite/news-next/src/workload-test.mjs]: changed to two imports

**Pull request merged**
